Kubernetes/legos/k8s_get_cluster_health/k8s_get_cluster_health.py

```diff
@@ -0,0 +1,127 @@
+    event_string = ''
+    try:
+        events = node_api.list_event_for_all_namespaces(field_selector=field_selector)
+
+        # Print the details of each event
+        for event in events.items:
+            event_string = event_string + f"Event: {event.metadata.name} - {event.type} - {event.reason} - {event.message} - {event.last_timestamp}" + '\n'
+        
+    except ApiException as e:
+        raise e 
+
+    return event_string
+
+def k8s_get_cluster_health(handle) -> Tuple:
+    """k8s_get_cluster_health This function takes the Handle as an input parameter,
+       finds out all the Nodes present in the Cluster, Finds out the following
+       * Any abnormal events seen on the node
+       * If the node is under pressure
+       * Node Status & Pod Status
+
+       :type handle: object
+       :param handle: Object returned from task.validator(...)
+
+       :rtype: Tuple
+       :param: Tuple that contains result and any errors   
+    """
+    node_api = pods_api = client.CoreV1Api(api_client=handle)
+
+    nodes = node_api.list_node() 
+    for node in nodes.items():
+        # Lets check Node Pressure, more than 80%, will need to to raise an exception
+        cpu_usage = normalize_cpu(node.status.allocatable['cpu'])
+        cpu_capacity = normalize_cpu(node.status.capacity['cpu'])
+        mem_usage = normalize_memory(node.status.allocatable['memory'])
+        mem_capacity = normalize_memory(node.status.capacity['memory'])
+        cpu_usage_percent = (cpu_usage / cpu_capacity) * 100
+        mem_usage_percent = (mem_usage / mem_capacity) * 100 
+        if cpu_usage_percent >= 80 or mem_usage_percent >= 80:
+            raise Exception(f"Node {node.metadata.name} Experiencing High CPU {round(cpu_usage_percent,2)}% / MEM {round(mem_usage_percent,2)}% usage")
+        
+        # Lets get abnormal events. Lets go with `warning` as the default level
+        events = k8s_get_abnormal_events(node_api, node.metadata.name)
+        if events != '':
+            return (False, [{"error": events}])
+        
+        # Get Node & Pod Condition
+        node_condition = ''
+        conditions = node.status.conditions
+        for condition in conditions:
+            if condition.type == 'Ready' and condition.status == 'True':
+                break
+            else:
+                node_condition += f"\tNode {node.metadata.name} is not ready" 
+        
+        if not node_condition:
+            return (False, [{"error": node_condition}])
+
+        # Check the status of the Kubernetes pods
+        pods = pods_api.list_pod_for_all_namespaces()
+        pod_condition = ''
+        for pod in pods.items:
+            conditions = pod.status.conditions
+            for condition in conditions:
+                if condition.type == 'Ready' and condition.status == 'True':
+                    break
+            else:
+                pod_condition += f"\tPod {pod.metadata.namespace}/{pod.metadata.name} is not ready"
```

Comment:
Such kind of check is not at all consumable by the remediation runbook. Please think about how this check output will be used.

Reply:
Modified to return Python dict with the key `unhealthy_pod` that has a list of dictionaries having details of the unhealthy pods.